File: Filewatcherd-TypeScript/src/lib/ChangedFileEntry.ts
/*******************************************************************************
* Copyright (c) 2019 IBM Corporation and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v2.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v20.html
*
* Contributors:
*     IBM Corporation - initial API and implementation
*******************************************************************************/

import * as we from "./WatchEventEntry";

/**
 * Simple representation of a single change: the file/dir path that changed,
 * what type of change, and when. These are then consumed by the batch
 * processing utility.
 */
export class ChangedFileEntry {
    private readonly _eventType: we.EventType;

    private readonly _timestamp: number;
    private readonly _path: string;

    private readonly _directory: boolean;

    constructor(path: string, type: we.EventType, timestamp: number, directory: boolean) {
        if (!path || !type || timestamp < 0) {
            throw new Error("Invalid parameter '" + path + "' '" + type + "' '" + timestamp + "'");
        }

        this._eventType = type;
        this._path = path;
        this._timestamp = timestamp;
        this._directory = directory;
    }

    public toJson(): IChangedFileEntryJson {

        const result: IChangedFileEntryJson = {
            directory : this._directory,
            path  : this._path,
            timestamp : this._timestamp,
            type : we.eventTypetoString(this._eventType),
        };

        return result;
    }

    public get timestamp(): number {
        return this._timestamp;
    }

    public get eventType(): we.EventType {
        return this._eventType;
    }

    public get path(): string {
        return this._path;
    }

    public get directory(): boolean {
        return this._directory;
    }
}

/** Represents a single file change that is then communicated and sent in the HTTP POST request. */
export interface IChangedFileEntryJson {
    path: string;
    timestamp: number;
    type: string;
    directory: boolean;
}

File: Tests/FilewatcherTests/src/org/eclipse/codewind/filewatchers/test/infrastructure/ChangedFileEntry.java
/*******************************************************************************
 * Copyright (c) 2020 IBM Corporation
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/

package org.eclipse.codewind.filewatchers.test.infrastructure;

import java.io.File;

import org.eclipse.codewind.filewatchers.test.json.ChangedFileEntryJson;

public class ChangedFileEntry {

	public static enum EventType {
		CREATE, MODIFY, DELETE
	}

	private final EventType eventType;

	private final long timestamp;

	private final String path;

	private final boolean directory;

	public ChangedFileEntry(ChangedFileEntryJson j) {
		this.timestamp = j.getTimestamp();
		this.path = j.getPath();
		this.directory = j.isDirectory();

		if (j.getType().equals("CREATE")) {
			this.eventType = EventType.CREATE;
		} else if (j.getType().equals("MODIFY")) {
			this.eventType = EventType.MODIFY;
		} else if (j.getType().equals("DELETE")) {
			this.eventType = EventType.DELETE;
		} else {
			throw new IllegalArgumentException();
		}
	}

	public ChangedFileEntry(EventType eventType, long timestamp, String path, boolean isDirectory) {
		this.eventType = eventType;
		this.timestamp = timestamp;
		this.path = path;
		this.directory = isDirectory;
	}

	public EventType getEventType() {
		return eventType;
	}

	public long getTimestamp() {
		return timestamp;
	}

	public String getPath() {
		return path;
	}

	public boolean isDirectory() {
		return directory;
	}

	public static File convertChangeEntryPathToAbsolute(File parent, String path) {

		// This is hack, but works fine here in the context of an automated test.
		if (path.startsWith("/home") || path.toLowerCase().startsWith("/users")
				|| path.toLowerCase().startsWith("c:\\users") || path.toLowerCase().startsWith("/tmp")) {
			return new File(path);
		}

		path = path.replace("/", File.separator);

		return new File(parent, path);

	}

	@Override
	public String toString() {

		return "[" + eventType.name() + "] " + path + " @ " + timestamp;
	}
}
